src/theory/uf/distinct_extension.h

```diff
@@ -0,0 +1,105 @@
+#include "theory/theory_model.h"
+#include "theory/theory_state.h"
+
+namespace cvc5::internal {
+namespace theory {
+namespace uf {
+
+class DistinctProofGenerator;
+
+/**
+ * Lazy handling of distinct. This subsolver manages distinct constraints by
+ * tracking if the equivalence class of two children of positively asserted
+ * distinct are merged. It also reduces negatively asserted distincts if
+ * necessary based on the constructed model.
+ */
+class DistinctExtension : protected EnvObj
+{
+ public:
+  DistinctExtension(Env& env, TheoryState& state, TheoryInferenceManager& im);
+  /** Check, called after the fact queue of the theory is processed. */
+  void check(Theory::Effort level);
+  /** Do we need a last call check? */
+  bool needsCheckLastEffort();
+  /** Assert distinct, may return a conflict */
+  void assertDistinct(TNode atom, bool pol, TNode fact);
+  /** Called when t1 and t2 merge, may return a conflict */
+  void eqNotifyMerge(TNode t1, TNode t2);
+
+ private:
+  Node d_false;
+  /** Reference to the state object */
+  TheoryState& d_state;
+  /** Reference to the inference manager */
+  TheoryInferenceManager& d_im;
+  /** maps nodes to an index in a vector */
+  using NodeUIntMap = context::CDHashMap<Node, size_t>;
+  // For lazy handling of distinct. We map equivalence classes to a set of
+  // distinct constraints that they occur in. For example, if we have an
+  // equivalence class {a, b} where a is the representative, and
+  // distinct(b,c,d) distinct(e,f,a) are constraints, then for example:
+  // d_ndistinct[a] = 2,
+  // d_eqcToDistinct[a] = {distinct(b,c,d), distinct(e,f,a)},
+  // d_eqcToDMem[a] = {b,a}
+  // Similarly if {d} is an equivalence class, then
+  // d_ndistinct[d] = 1,
+  // d_eqcToDistinct[d] = {distinct(b,c,d)},
+  // d_eqcToDMem[d] = {d}
+  // If {a,b} and {d} merge, we recognize a conflict since distinct(b,c,d) is
+  // in both lists, where b=d is the explanation for the conflict, accessible
+  // via d_eqcToDMem.
+  /** The number of entries in d_eqcToDistinct/d_eqcToMem that are valid */
+  NodeUIntMap d_ndistinct;
+  /** Mapping from equivalence classes to the list of distincts that they belong
+   * to */
+  std::map<Node, std::vector<Node>> d_eqcToDistinct;
+  /** The corresponding term in the equivalence class, for each entry in
+   * d_eqcToDistinct */
+  std::map<Node, std::vector<Node>> d_eqcToDMem;
+  /** The set of asserted negated distinct constraints */
+  context::CDList<Node> d_negDistinct;
+  /** The number of constraints in the above list we have reduced. */
+  context::CDO<size_t> d_negDistinctIndex;
+  /** The set of asserted negated distinct constraints */
+  context::CDList<Node> d_posDistinct;
+  /** A proof generator for disequal congruent terms */
+  std::shared_ptr<DistinctProofGenerator> d_dproof;
+  /** Eager proof generator */
+  std::shared_ptr<EagerProofGenerator> d_epg;
+  /** The pending conflict if one exists */
+  context::CDO<Node> d_pendingConflict;
```

Comment:
not sure if here but somewhere it should be clarified that pending conflicts are always of a given form.